pkg/apis/admissionregistration/validation/validation.go

```diff
@@ -1328,3 +1382,243 @@ func isCELIdentifier(name string) bool {
+	trimmedExpression := strings.TrimSpace(applyConfig.Expression)
+	if len(trimmedExpression) == 0 {
+		allErrors = append(allErrors, field.Required(fldPath.Child("expression"), ""))
+	} else {
+
+		envType := environment.NewExpressions
+		if opts.preexistingExpressions.applyConfigurationExpressions.Has(applyConfig.Expression) {
+			envType = environment.StoredExpressions
+		}
+		accessor := &mutating.ApplyConfigurationCondition{
+			Expression: trimmedExpression,
+		}
+		opts := plugincel.OptionalVariableDeclarations{HasParams: paramKind != nil, HasAuthorizer: true, StrictCost: true, HasObjectTypes: true}
+		result := compiler.CompileCELExpression(accessor, opts, envType)
+
+		if result.Error != nil {
+			allErrors = append(allErrors, convertCELErrorToValidationError(fldPath.Child("expression"), accessor, result.Error))
+		}
+	}
+	return allErrors
+}
+
+func validateJSONPatch(compiler plugincel.Compiler, jsonPatch admissionregistration.JSONPatch, paramKind *admissionregistration.ParamKind, opts validationOptions, fldPath *field.Path) (allErrors field.ErrorList) {
+	for i, patch := range jsonPatch {
+		trimmedValueExpression := strings.TrimSpace(patch.ValueExpression)
+		hasValue := len(trimmedValueExpression) > 0
+		hasFrom := len(patch.From) > 0
+		needValue := false
+		needFrom := false
+		switch patch.Op {
+		case admissionregistration.Add, admissionregistration.Replace, admissionregistration.Test:
+			needValue = true
+		case admissionregistration.Move, admissionregistration.Copy:
+			needFrom = true
+		case admissionregistration.Remove:
+		default:
+			allErrors = append(allErrors, field.NotSupported(fldPath.Index(i).Child("op"), patch.Op, supportedJSONPatchOperation.List()))
+		}
+
+		if len(patch.Path) == 0 {
+			allErrors = append(allErrors, field.Required(fldPath.Index(i).Child("path"), ""))
+		} else {
+			allErrors = append(allErrors, validateJSONPointer(patch.Path, fldPath.Child("path"))...)
+		}
+
+		if needValue && !hasValue {
+			allErrors = append(allErrors, field.Required(fldPath.Index(i).Child("valueExpression"), fmt.Sprintf("must be specified when op is %s", string(patch.Op))))
+		}
+		if !needValue && hasValue {
+			allErrors = append(allErrors, field.Invalid(fldPath.Index(i).Child("valueExpression"), "{valueExpression}", fmt.Sprintf("must not be specified when op is %s", string(patch.Op))))
+		}
+		if needValue && hasValue {
+			envType := environment.NewExpressions
+			if opts.preexistingExpressions.jsonPatchExpressions.Has(patch.ValueExpression) {
+				envType = environment.StoredExpressions
+			}
+			accessor := &mutating.JSONPatchCondition{
+				ValueExpression: trimmedValueExpression,
+			}
+			result := compiler.CompileCELExpression(accessor, plugincel.OptionalVariableDeclarations{HasParams: paramKind != nil, HasAuthorizer: true, StrictCost: true}, envType)
```

Comment:
Do we have any way to try to guess whether we'll get a valid yaml object after evaluating the expression?  Even if we cannot check that types are correct, can we take a guess at syntactic validity?  Even a warning if we can get it mostly right could help.

Reply:
Good question.  All CEL data literal declarations has maps, arrays and json compatible scalar types.  But we do have a few special types floating around like authorizer...   I'll see if there is anything we can do in the type checker to detect these.

Reply:
I looked into this.  There's not a lot we more we can do at compile time.  There are a few constraints here.  We don't always know what resource is being mutated statically, and we don't know the type of what the json patch path refers to, so we can't do any type matching at those levels.  Because of that, we're telling the CEL compiler that the return type is `cel.DynType`  and letting the expression return most anything.  We really want a subset of DynType that's just types that map to JSON values, but there's nothing in the type system available to achieve that.  The good news is that MOST types do map to JSON values, so there is a relatively small set of special types that will fail at runtime, but I don't think we can detect them.